lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,410 @@
+///   (_props) {
+///     final defaultProps = useMemo(() {
+///       return WithTransitionExample()..initiallyShown = false;
+///     }, const []);
+///     final props = defaultProps..addProps(_props);
+///
+///     final isShown = useState(props.initiallyShown);
+///
+///     final classes = ClassNameBuilder.fromProps(props)
+///       ..add('fade')
+///       ..add('in', isShown.value);
+///
+///     return Dom.div()(
+///       Dom.div()(
+///         (Dom.button()
+///           ..type = 'button'
+///           ..onClick = (useCallback((_) {
+///             isShown.setWithUpdater((isShown) => !isShown);
+///           }, const []))
+///         )('Toggle Visibility'),
+///       ),
+///       (WithTransition()
+///         ..isShown = isShown.value
+///         ..onWillShow = () {
+///           // Do something that should happen before the CSS transition completes.
+///         }
+///         ..onDidShow = () {
+///           // Do something that shouldn't happen until the CSS transition completes.
+///         }
+///         ..onWillHide = () {
+///           // Do something that should happen before the CSS transition completes.
+///         }
+///         ..onDidHide = () {
+///           // Do something that shouldn't happen until the CSS transition completes.
+///         }
+///       )(
+///         (Dom.div()
+///           ..className = classes.toClassName()
+///         )(
+///           // Some children...
+///         ),
+///       )
+///     );
+///   },
+///   $WithTransitionExampleConfig, // ignore: undefined_identifier
+/// );
+/// ```
+///
+/// __Wrapping around custom components__
+///
+/// If you want to wrap around another component instead of around a single `Dom` element child as shown in the
+/// first example above, you must ensure that the value of `props.ref` is forwarded to the `Element` that has
+/// the CSS transition using `uiForwardRef`:
+///
+/// ```dart
+/// import 'package:over_react/over_react.dart';
+///
+/// mixin CustomChildProps on UiProps {}
+///
+/// UiFactory<CustomChildProps> CustomChild = uiForwardRef(
+///   (props, ref) {
+///     return (Dom.div()..ref = ref)(
+///       props.children,
+///     ),
+///   },
+///   $CustomChildConfig, // ignore: undefined_identifier
+/// );
+/// ```
+UiFactory<WithTransitionProps> WithTransition = memo(uiFunction(
+  (_props) {
+    // ----- Default Props ----- //
+    final defaultProps = useMemo(() {
+      return WithTransition()
+      ..isShown = false
+      ..transitionTimeout = const Duration(seconds: 1)
+      ..childPropsByPhase = const {
+        v2.TransitionPhase.PRE_SHOWING: {},
+        v2.TransitionPhase.SHOWING: {},
+        v2.TransitionPhase.SHOWN: {},
+        v2.TransitionPhase.HIDING: {},
+        v2.TransitionPhase.HIDDEN: {},
+      }
+      ..addProps(v2.TransitionPropsMixin.defaultProps);
+    }, const []);
+    final props = defaultProps..addProps(_props);
```

Comment:
#nit prefer using null-awares to default props like mentioned above

Reply:
Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8